### PR TITLE
Implement projection for Query

### PR DESCRIPTION
**Pull request by YazhiFeng**
Projection has been implemented and a test case is added.

### CODE REVIEW

File: src/k2/module/k23si/Module.cpp [hunk 1]
```diff
@@ -302,8 +302,17 @@ K23SIPartitionModule::handleQuery(dto::K23SIQueryRequest&& request, dto::K23SIQu
...
+                // TODO apply filter 
+                // apply projection if the user call addProjection
+                if (request.projection.size() == 0) {
+                    // want all fields 
+                    response.results.push_back(viter->value.share());
+                } else {
+                    // serialize partial SKVRecord according to projection
+                    dto::SKVRecord::Storage storage;
+                    _makeProjection(viter->value, request, storage);
+                    response.results.push_back(storage.share());
```
Comment (jfunston) [inline src/k2/module/k23si/Module.cpp:314]: You can use std::move(storage) instead of share()

File: src/k2/module/k23si/Module.cpp [hunk 2]
```diff
@@ -784,6 +793,38 @@ bool K23SIPartitionModule::_parsePartialRecord(dto::K23SIPartialUpdateRequest& r
...
+            // write field value into payload
+            _copyPayloadBaseToUpdate(fullRec.fieldData, payload, schema.fields[i].type);
+            excludedFields[i] = false;
+        }
+    }
+    
+    // set cursor(0) of base payload
+    fullRec.fieldData.seek(0);
+
+    projectionRec.excludedFields = excludedFields;
```
Comment (jfunston) [inline src/k2/module/k23si/Module.cpp:821]: You can use std::move here too

File: test/k23si/QueryTest.cpp [hunk 3]
```diff
@@ -120,14 +121,15 @@ class QueryTest {
 
 seastar::future<> doQuery(const k2::String& start, const k2::String& end, int32_t limit, bool reverse, 
                           uint32_t expectedRecords, uint32_t expectedPaginations, 
-                          k2::Status expectedStatus=k2::dto::K23SIStatus::OK) {
+                          k2::Status expectedStatus=k2::dto::K23SIStatus::OK,
+                          std::vector<k2::String> projection=std::vector<k2::String>()) {
```
Comment (jfunston) [inline test/k23si/QueryTest.cpp:125]: I think you should change this function so it returns the result set to the caller, then the tests for projection can test that the correct data is returned. The already existing tests can just ignore the result.

File: test/k23si/QueryTest.cpp [hunk 4]
```diff
@@ -248,10 +273,18 @@ seastar::future<> runScenario01() {
     });
 }
 
-// Query conflict cases
+// Query projection cases
 seastar::future<> runScenario02() {
     K2INFO("runScenario02");
 
+    K2INFO("Projection reads for giving fields");
+    return doQuery("a", "c", -1, false, 2, 1, k2::dto::K23SIStatus::OK, {"partition"});
```
Comment (jfunston) [inline test/k23si/QueryTest.cpp:281]: Like I mentioned above, let's check that the data is correct. Also, can you add a test that tries to project a field that is not part of the schema?

**Review by jfunston — CHANGES_REQUESTED** [review]

**Comment by YazhiFeng** [comment]
Use std::move to return value, func doQuery() returns val_set, and add a testcase.

**Review by jfunston — APPROVED** [review]
Looks good, thanks. You can start on https://github.com/futurewei-cloud/chogori-platform/issues/59 next. As always feel free to ask me questions on the issue or in the open source meeting.

File: src/k2/module/k23si/Module.cpp [hunk 5]
```diff
@@ -784,6 +793,38 @@ bool K23SIPartitionModule::_parsePartialRecord(dto::K23SIWriteRequest& request,
     return true;
 }
 
+void K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto::K23SIQueryRequest& request, 
+        dto::SKVRecord::Storage& projectionRec) {
+    auto schemaIt = _schemas.find(request.key.schemaName);
+    auto schemaVer = schemaIt->second.find(fullRec.schemaVersion);
+    dto::Schema& schema = schemaVer->second;
+    std::vector<bool> excludedFields(schema.fields.size(), true);   // excludedFields for projection
+    Payload payload(Payload::DefaultAllocator);                     // payload for projection
```
Comment (jerryhfeng) [inline src/k2/module/k23si/Module.cpp:802]: Minor: Could you rename it to projectedPayload for better readability.

**Pull request merged**